File: src/definitions.ts
import { Color } from "./color";

// Colors

const none = new Color("#000000").setAlpha(0);
const white = new Color("#ffffff");
const black = new Color("#000000");
const grey = new Color("#bec9d3", "#2e2e37");
const greyStrong = new Color("#6b7a88", "#a8a8b1");
const greyLight = new Color("#e7ecf2", "#1e1d27");
const greyOnBlack = new Color("#65696e", grey);

const purple = new Color("#7d46fc");
const blue = new Color("#004bff");
const cyan = new Color("#00d2ff");
const green = new Color("#00ff68");
const yellow = new Color("#ffca00");
const orange = new Color("#ff9900");
const red = new Color("#ff0032");

const main = new Color(white, black);
const mainNegative = new Color(black, white);
const mainContrast = new Color("#f4f6fc", "#0a0b0f");
const mainContrastTransparent = new Color("#99a7e1", "#616a93").setAlpha(0.1);

const primary = new Color(blue, purple);
const primaryNegative = white;
const primaryHighlight = new Color("#0064ff", primary).setAlpha(0.15, 0.25);
const primaryHighlightStrong = primaryHighlight.withAlpha(0.3, 0.5);
const primaryHighlightLight = primaryHighlight.withAlpha(0.05, 0.08);

const added = green;
const addedHighlight = added.withAlpha(0.1, 0.2);
const deleted = red;
const deletedHighlight = deleted.withAlpha(0.1, 0.2);

const error = red;
const warning = orange;

const find = yellow.withAlpha(0.6);
const findHighlight = find.withAlpha(0.3);

const border = greyLight;

// Theme settings

export default {
    colors: {
        "activityBar.background": main,
        "activityBar.border": none,
        "activityBar.foreground": grey,
        "activityBarBadge.background": mainNegative,
        "activityBarBadge.foreground": main,
        "badge.background": mainNegative,
        "badge.foreground": main,
        "button.background": primary,
        "button.foreground": primaryNegative,
        "contrastBorder": none,
        "diffEditor.insertedTextBackground": addedHighlight,
        "diffEditor.removedTextBackground": deletedHighlight,
        "dropdown.background": mainContrast,
        "dropdown.border": border,
        "dropdown.foreground": mainNegative,
        "editor.background": main,
        "editor.findMatchBackground": find,
        "editor.findMatchHighlightBackground": findHighlight,
        "editor.foreground": mainNegative,
        "editor.lineHighlightBackground": primaryHighlightLight,
        "editor.lineHighlightBorder": none,
        "editor.selectionBackground": primaryHighlightStrong,
        "editor.selectionForeground": main,
        "editor.wordHighlightBackground": primaryHighlight,
        "editor.wordHighlightStrongBackground": primaryHighlightStrong,
        "editorBracketMatch.background": primaryHighlightStrong,
        "editorBracketMatch.border": none,
        "editorCursor.foreground": primary,
        "editorError.foreground": error,
        "editorGroup.border": mainContrast,
        "editorGroupHeader.tabsBackground": main,
        "editorGroupHeader.tabsBorder": none,
        "editorGutter.addedBackground": added,
        "editorGutter.deletedBackground": deleted,
        "editorGutter.modifiedBackground": mainNegative,
        "editorIndentGuide.background": mainContrast,
        "editorLineNumber.foreground": new Color("#ebf0f5", greyLight),
        "editorLink.activeForeground": primary,
        "editorOverviewRuler.addedForeground": none,
        "editorOverviewRuler.border": none,
        "editorOverviewRuler.bracketMatchForeground": primaryHighlightStrong,
        "editorOverviewRuler.errorForeground": error,
        "editorOverviewRuler.findMatchForeground": find,
        "editorOverviewRuler.modifiedForeground": mainNegative,
        "editorOverviewRuler.selectionHighlightForeground": primaryHighlightStrong,
        "editorOverviewRuler.warningForeground": warning,
        "editorOverviewRuler.wordHighlightForeground": primaryHighlight,
        "editorOverviewRuler.wordHighlightStrongForeground": primaryHighlightStrong,
        "editorRuler.foreground": mainContrast,
        "editorSuggestWidget.foreground": mainNegative,
        "editorWarning.foreground": warning,
        "editorWidget.background": mainContrast,
        "editorWidget.border": border,
        "errorForeground": error,
        "extensionButton.prominentBackground": primary,
        "extensionButton.prominentForeground": primaryNegative,
        "extensionButton.prominentHoverBackground": primary,
        "focusBorder": border,
        "foreground": mainNegative,
        "gitDecoration.ignoredResourceForeground": grey,
        "gitDecoration.modifiedResourceForeground": mainNegative,
        "gitDecoration.untrackedResourceForeground": mainNegative,
        "input.background": mainContrast,
        "input.border": border,
        "input.foreground": mainNegative,
        "input.placeholderForeground": grey,
        "list.activeSelectionBackground": mainNegative,
        "list.activeSelectionForeground": main,
        "list.dropBackground": primaryHighlightStrong,
        "list.focusBackground": primaryHighlight,
        "list.focusForeground": mainNegative,
        "list.highlightForeground": mainNegative,
        "list.hoverBackground": mainContrast,
        "list.inactiveSelectionBackground": mainContrast,
        "list.inactiveSelectionForeground": mainNegative,
        "panel.background": mainContrast,
        "panel.border": none,
        "panelTitle.activeBorder": grey,
        "peekView.border": mainNegative,
        "peekViewEditor.background": none,
        "peekViewEditor.matchHighlightBackground": find,
        "peekViewResult.background": none,
        "peekViewResult.matchHighlightBackground": find,
        "peekViewResult.selectionBackground": mainNegative,
        "peekViewResult.selectionForeground": main,
        "peekViewTitle.background": none,
        "peekViewTitleDescription.foreground": grey,
        "peekViewTitleLabel.foreground": mainNegative,
        "progressBar.background": primary,
        "scrollbar.shadow": none,
        "scrollbarSlider.background": mainContrastTransparent,
        "scrollbarSlider.hoverBackground": mainNegative,
        "sideBar.background": main,
        "sideBar.foreground": greyStrong,
        "sideBarSectionHeader.background": mainContrast,
        "sideBarSectionHeader.foreground": greyStrong,
        "sideBarTitle.foreground": grey,
        "statusBar.background": black,
        "statusBar.foreground": greyOnBlack,
        "statusBar.noFolderBackground": black,
        "tab.activeBackground": mainNegative,
        "tab.activeForeground": main,
        "tab.border": none,
        "tab.inactiveBackground": none,
        "tab.inactiveForeground": grey,
        "terminal.ansiBlack": mainNegative,
        "terminal.ansiBlue": blue,
        "terminal.ansiBrightBlack": mainNegative,
        "terminal.ansiBrightBlue": blue,
        "terminal.ansiBrightCyan": cyan,
        "terminal.ansiBrightGreen": green,
        "terminal.ansiBrightMagenta": purple,
        "terminal.ansiBrightRed": red,
        "terminal.ansiBrightWhite": greyStrong,
        "terminal.ansiBrightYellow": yellow,
        "terminal.ansiCyan": cyan,
        "terminal.ansiGreen": green,
        "terminal.ansiMagenta": purple,
        "terminal.ansiRed": red,
        "terminal.ansiWhite": greyStrong,
        "terminal.ansiYellow": yellow,
        "terminal.foreground": mainNegative,
        "terminal.selectionBackground": primaryHighlightStrong,
        "terminalCursor.foreground": primary,
        "textLink.activeForeground": primary,
        "textLink.foreground": primary,
        "titleBar.activeBackground": main,
        "titleBar.activeForeground": greyStrong,
        "titleBar.inactiveBackground": main,
        "titleBar.inactiveForeground": grey,
        "welcomePage.buttonBackground": mainContrast,
        "welcomePage.buttonHoverBackground": primaryHighlight,
        "widget.shadow": none,
    },
    tokenColors: [
        {
            scope: ["comment", "string.quoted.docstring"],
            settings: {
                foreground: grey,
            },
        },
        {
            scope: ["string"],
            settings: {
                foreground: greyStrong,
            },
        },
        {
            scope: [
                "punctuation.definition.string",
                "storage.type.string.python",
            ],
            settings: {
                foreground: mainNegative,
            },
        },
        {
            scope: [
                "beginning.punctuation",
                "entity.name.section.group-title",
                "entity.name.tag",
                "entity.other.attribute-name.class",
                "entity.other.attribute-name.id",
                "keyword.const",
                "keyword.control",
                "keyword.function",
                "keyword.import",
                "keyword.operator.assignment",
                "keyword.operator.comparison",
                "keyword.operator.decrement",
                "keyword.operator.expression",
                "keyword.operator.increment",
                "keyword.operator.increment-decrement",
                "keyword.operator.logical",
                "keyword.operator.new",
                "keyword.operator.other",
                "keyword.other.special-method",
                "keyword.package",
                "keyword.type",
                "keyword.var",
                "markup.heading",
                "meta.function.decorator",
                "meta.tag.sgml.doctype.html",
                "punctuation.separator.key-value",
                "storage.modifier",
                "storage.type.class",
                "storage.type.enum",
                "storage.type.function",
                "storage.type.import",
                "storage.type.interface",
                "storage.type.js",
                "storage.type.namespace",
                "storage.type.property",
                "storage.type.string.python",
                "storage.type.ts",
            ],
            settings: {
                fontStyle: "bold",
            },
        },
    ],
};

File: definitions.py
"""Themes color definitions."""

from build import ColorSplit


# Colors

colorNone = ColorSplit('#000000').set_alpha(0)
colorWhite = ColorSplit('#ffffff')
colorBlack = ColorSplit('#000000')
colorGrey = ColorSplit('#bec9d3', '#2e2e37')
colorGreyStrong = ColorSplit('#6b7a88', '#a8a8b1')
colorGreyLight = ColorSplit('#e7ecf2', '#1e1d27')
colorGreyOnBlack = ColorSplit('#65696e', colorGrey)

colorPurple = ColorSplit('#7d46fc')
colorBlue = ColorSplit('#004bff')
colorCyan = ColorSplit('#00d2ff')
colorGreen = ColorSplit('#00ff68')
colorYellow = ColorSplit('#ffca00')
colorOrange = ColorSplit('#ff9900')
colorRed = ColorSplit('#ff0032')

colorMain = ColorSplit(colorWhite, colorBlack)
colorMainNegative = ColorSplit(colorBlack, colorWhite)
colorMainContrast = ColorSplit('#f4f6fc', '#0a0b0f')
colorMainContrastTransparent = ColorSplit('#99a7e1', '#616a93').set_alpha(0.1)

colorPrimary = ColorSplit(colorBlue, colorPurple)
colorPrimaryNegative = colorWhite
colorPrimaryHighlight = ColorSplit(
    '#0064ff', colorPrimary).set_alpha(0.15, 0.25)
colorPrimaryHighlightStrong = colorPrimaryHighlight.with_alpha(0.3, 0.5)
colorPrimaryHighlightLight = colorPrimaryHighlight.with_alpha(0.05, 0.08)

colorAdded = colorGreen
colorAddedHighlight = colorAdded.with_alpha(0.1, 0.2)
colorDeleted = colorRed
colorDeletedHighlight = colorDeleted.with_alpha(0.1, 0.2)

colorError = colorRed
colorWarning = colorOrange

colorFind = colorYellow.with_alpha(0.6)
colorFindHighlight = colorFind.with_alpha(0.3)

colorBorder = colorGreyLight


# Theme definitions

definitions = {
    'colors': {
        'activityBar.background': colorMain,
        'activityBar.border': colorNone,
        'activityBar.foreground': colorGrey,
        'activityBarBadge.background': colorMainNegative,
        'activityBarBadge.foreground': colorMain,
        'badge.background': colorMainNegative,
        'badge.foreground': colorMain,
        'button.background': colorPrimary,
        'button.foreground': colorPrimaryNegative,
        'contrastBorder': colorNone,
        'diffEditor.insertedTextBackground': colorAddedHighlight,
        'diffEditor.removedTextBackground': colorDeletedHighlight,
        'dropdown.background': colorMainContrast,
        'dropdown.border': colorBorder,
        'dropdown.foreground': colorMainNegative,
        'editor.background': colorMain,
        'editor.findMatchBackground': colorFind,
        'editor.findMatchHighlightBackground': colorFindHighlight,
        'editor.foreground': colorMainNegative,
        'editor.lineHighlightBackground': colorPrimaryHighlightLight,
        'editor.lineHighlightBorder': colorNone,
        'editor.selectionBackground': colorPrimaryHighlightStrong,
        'editor.selectionForeground': colorMain,
        "editor.selectionHighlightBorder":
            ColorSplit(colorNone, colorPrimaryHighlight),
        'editor.wordHighlightBackground': colorPrimaryHighlight,
        'editor.wordHighlightStrongBackground': colorPrimaryHighlightStrong,
        'editorBracketMatch.background': colorPrimaryHighlightStrong,
        'editorBracketMatch.border': colorNone,
        'editorCursor.foreground': colorPrimary,
        'editorError.foreground': colorError,
        'editorGroup.border': colorMainContrast,
        'editorGroupHeader.tabsBackground': colorMain,
        'editorGroupHeader.tabsBorder': colorNone,
        'editorGutter.addedBackground': colorAdded,
        'editorGutter.deletedBackground': colorDeleted,
        'editorGutter.modifiedBackground': colorMainNegative,
        'editorIndentGuide.background': colorMainContrast,
        'editorLineNumber.foreground': ColorSplit('#ebf0f5', colorGreyLight),
        'editorLink.activeForeground': colorPrimary,
        'editorOverviewRuler.addedForeground': colorNone,
        'editorOverviewRuler.border': colorNone,
        'editorOverviewRuler.bracketMatchForeground':
            colorPrimaryHighlightStrong,
        'editorOverviewRuler.errorForeground': colorError,
        'editorOverviewRuler.findMatchForeground': colorFind,
        'editorOverviewRuler.modifiedForeground': colorMainNegative,
        'editorOverviewRuler.selectionHighlightForeground':
            colorPrimaryHighlightStrong,
        'editorOverviewRuler.warningForeground': colorWarning,
        'editorOverviewRuler.wordHighlightForeground': colorPrimaryHighlight,
        'editorOverviewRuler.wordHighlightStrongForeground':
            colorPrimaryHighlightStrong,
        'editorRuler.foreground': colorMainContrast,
        'editorSuggestWidget.foreground': colorMainNegative,
        'editorWarning.foreground': colorWarning,
        'editorWidget.background': colorMainContrast,
        'editorWidget.border': colorBorder,
        'errorForeground': colorError,
        'extensionButton.prominentBackground': colorPrimary,
        'extensionButton.prominentForeground': colorPrimaryNegative,
        'extensionButton.prominentHoverBackground': colorPrimary,
        'focusBorder': colorBorder,
        'foreground': colorMainNegative,
        'gitDecoration.ignoredResourceForeground': colorGrey,
        'gitDecoration.modifiedResourceForeground': colorMainNegative,
        'gitDecoration.untrackedResourceForeground': colorMainNegative,
        'input.background': colorMainContrast,
        'input.border': colorBorder,
        'input.foreground': colorMainNegative,
        'input.placeholderForeground': colorGrey,
        'list.activeSelectionBackground': colorMainNegative,
        'list.activeSelectionForeground': colorMain,
        'list.dropBackground': colorPrimaryHighlightStrong,
        'list.focusBackground': colorPrimaryHighlight,
        'list.focusForeground': colorMainNegative,
        'list.highlightForeground': colorMainNegative,
        'list.hoverBackground': colorMainContrast,
        'list.inactiveSelectionBackground': colorMainContrast,
        'list.inactiveSelectionForeground': colorMainNegative,
        'panel.background': colorMainContrast,
        'panel.border': colorNone,
        'panelTitle.activeBorder': colorGrey,
        'peekView.border': colorMainNegative,
        'peekViewEditor.background': colorNone,
        'peekViewEditor.matchHighlightBackground': colorFind,
        'peekViewResult.background': colorNone,
        'peekViewResult.fileForeground': colorGreyStrong,
        'peekViewResult.lineForeground': colorGreyStrong,
        'peekViewResult.matchHighlightBackground': colorFind,
        'peekViewResult.selectionBackground': colorMainNegative,
        'peekViewResult.selectionForeground': colorMain,
        'peekViewTitle.background': colorNone,
        'peekViewTitleDescription.foreground': colorGrey,
        'peekViewTitleLabel.foreground': colorMainNegative,
        'progressBar.background': colorPrimary,
        'scrollbar.shadow': colorNone,
        'scrollbarSlider.background': colorMainContrastTransparent,
        'scrollbarSlider.hoverBackground': colorMainNegative,
        'sideBar.background': colorMain,
        'sideBar.foreground': colorGreyStrong,
        'sideBarSectionHeader.background': colorMainContrast,
        'sideBarSectionHeader.foreground': colorGreyStrong,
        'sideBarTitle.foreground': colorGrey,
        'statusBar.background': colorBlack,
        'statusBar.foreground': colorGreyOnBlack,
        'statusBar.noFolderBackground': colorBlack,
        'tab.activeBackground': colorMainNegative,
        'tab.activeForeground': colorMain,
        'tab.border': colorNone,
        'tab.inactiveBackground': colorNone,
        'tab.inactiveForeground': colorGrey,
        'terminal.ansiBlack': colorMainNegative,
        'terminal.ansiBlue': colorBlue,
        'terminal.ansiBrightBlack': colorMainNegative,
        'terminal.ansiBrightBlue': colorBlue,
        'terminal.ansiBrightCyan': colorCyan,
        'terminal.ansiBrightGreen': colorGreen,
        'terminal.ansiBrightMagenta': colorPurple,
        'terminal.ansiBrightRed': colorRed,
        'terminal.ansiBrightWhite': colorWhite,
        'terminal.ansiBrightYellow': colorYellow,
        'terminal.ansiCyan': colorCyan,
        'terminal.ansiGreen': colorGreen,
        'terminal.ansiMagenta': colorPurple,
        'terminal.ansiRed': colorRed,
        'terminal.ansiWhite': colorWhite,
        'terminal.ansiYellow': colorYellow,
        'terminal.foreground': colorMainNegative,
        'terminal.selectionBackground': colorPrimaryHighlightStrong,
        'terminalCursor.foreground': colorPrimary,
        'textLink.activeForeground': colorPrimary,
        'textLink.foreground': colorPrimary,
        'titleBar.activeBackground': colorMain,
        'titleBar.activeForeground': colorGreyStrong,
        'titleBar.inactiveBackground': colorMain,
        'titleBar.inactiveForeground': colorGrey,
        'welcomePage.buttonBackground': colorMainContrast,
        'welcomePage.buttonHoverBackground': colorPrimaryHighlight,
        'widget.shadow': colorNone,
    },
    'tokenColors': [
        {
            'scope': [
                'comment',
                'string.quoted.docstring'
            ],
            'settings': {
                'foreground': colorGrey,
            },
        },
        {
            'scope': [
                'string'
            ],
            'settings': {
                'foreground': colorGreyStrong,
            },
        },
        {
            'scope': [
                'punctuation.definition.string',
                'storage.type.string.python',
            ],
            'settings': {
                'foreground': colorMainNegative,
            },
        },
        {
            'scope': [
                'beginning.punctuation',
                'entity.name.section.group-title',
                'entity.name.tag',
                'entity.other.attribute-name.class',
                'entity.other.attribute-name.id',
                'keyword.const',
                'keyword.control',
                'keyword.function',
                'keyword.import',
                'keyword.operator.assignment',
                'keyword.operator.comparison',
                'keyword.operator.decrement',
                'keyword.operator.expression',
                'keyword.operator.increment',
                'keyword.operator.increment-decrement',
                'keyword.operator.logical',
                'keyword.operator.new',
                'keyword.operator.other',
                'keyword.other.special-method',
                'keyword.package',
                'keyword.type',
                'keyword.var',
                'markup.heading',
                'meta.tag.sgml.doctype.html',
                'punctuation.separator.key-value',
                'storage.modifier',
                'storage.type.class',
                'storage.type.enum',
                'storage.type.function',
                'storage.type.import',
                'storage.type.interface',
                'storage.type.js',
                'storage.type.namespace',
                'storage.type.property',
                'storage.type.string.python',
                'storage.type.ts',
            ],
            'settings': {
                'fontStyle': 'bold',
            },
        },
    ],
}
